src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceParameterService.cs

```diff
@@ -0,0 +1,153 @@
+         IMethodSymbol methodSymbol, CancellationToken cancellationToken)
+        {
+            var mappingDictionary = TieExpressionToParameters(expression, methodSymbol);
+            expression = expression.TrackNodes(mappingDictionary.Values);
+            var identifiers = expression.DescendantNodes().Where(node => node is IdentifierNameSyntax);
+
+            if (!callSites.Keys.Any())
+            {
+                return null;
+            }
+
+            var firstCallSite = callSites.Keys.First();
+
+            var currentSolution = firstCallSite.Project.Solution;
+            foreach (var keyValuePair in callSites)
+            {
+                var document = currentSolution.GetDocument(keyValuePair.Key.Id);
+                var invocationExpressionList = keyValuePair.Value;
+                var generator = SyntaxGenerator.GetGenerator(document);
+                var semanticFacts = document.GetRequiredLanguageService<ISemanticFactsService>();
+                var invocationSemanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                var editor = new SyntaxEditor(await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false), generator);
+
+                foreach (var invocationExpression in invocationExpressionList)
+                {
+                    var newArgumentExpression = expression;
+                    var invocationArguments = invocationExpression.ArgumentList.Arguments;
+                    foreach (var argument in invocationArguments)
+                    {
+                        var associatedParameter = semanticFacts.FindParameterForArgument(invocationSemanticModel, argument, cancellationToken);
+                        var parenthesizedArgumentExpression = generator.AddParentheses(argument.Expression, false);
+                        if (!mappingDictionary.TryGetValue(associatedParameter, out var value))
+                        {
+                            continue;
+                        }
+
+                        newArgumentExpression = newArgumentExpression.ReplaceNode(newArgumentExpression.GetCurrentNode(value), parenthesizedArgumentExpression);
+                    }
+                    var allArguments = invocationExpression.ArgumentList.Arguments.Add(SyntaxFactory.Argument(newArgumentExpression.WithoutAnnotations(ExpressionAnnotationKind).WithAdditionalAnnotations(Simplifier.Annotation)));
+                    editor.ReplaceNode(invocationExpression, editor.Generator.InvocationExpression(invocationExpression.Expression, allArguments));
+                }
+
+                var newRoot = editor.GetChangedRoot();
+                document = document.WithSyntaxRoot(newRoot);
+                currentSolution = document.Project.Solution;
+            }
+
+            return currentSolution;
+        }
```

Comment:
this logic seems like it would be common to VB and C#, i don't really know why it's in teh subclass.

Reply:
The reason for it was that it was accessing specific SyntaxNode elements, like getting the arguments from the invocation expression syntax. Is it possible to move that into the abstract class?